pkg/installmanager/installmanager.go

```diff
@@ -788,6 +842,49 @@ func (m *InstallManager) generateAssets(cd *hivev1.ClusterDeployment) error {
 	return nil
 }
 
+// patchWorkerMachineSetManifest accepts a yaml manifest as []byte and patches the manifest to include an additional
+// security group filter if that manifest is the definition of a worker MachineSet. The security group is obtained from
+// an annotation (constants.ExtraWorkerSecurityGroupAnnotation) on the provided MachinePool and is the value of the
+// annotation (singular).
+func patchWorkerMachineSetManifest(manifestBytes []byte, pool *hivev1.MachinePool, logger log.FieldLogger) (modifiedBytes []byte, modified bool, err error) {
+	c, err := yamlutils.Decode(manifestBytes)
+	// Decoding the yaml here shouldn't produce an error. An error indicates that the
+	// installer produced yaml that could not be decoded.
+	if err != nil {
+		logger.WithError(err).Error("unable to decode manifest bytes")
+		return modifiedBytes, modified, err
+	}
+
+	// Return if file is not a worker MachineSet manifest
+	isMachineSet, _ := yamlutils.Test(c, "/kind", "MachineSet")
+	// Note: We have to escape the `/` in the label key as `~1` per https://datatracker.ietf.org/doc/html/rfc6901#section-3
+	isWorkerMachineSet, _ := yamlutils.Test(c, "/spec/template/metadata/labels/machine.openshift.io~1cluster-api-machine-type", "worker")
+	if !isMachineSet || !isWorkerMachineSet {
+		return modifiedBytes, modified, nil
+	}
+
+	securityGroupFilter := interface{}(pool.Annotations[constants.ExtraWorkerSecurityGroupAnnotation])
+	ops := yamlpatch.Patch{
+		yamlpatch.Operation{
+			Op:    "add",
+			Path:  yamlpatch.OpPath("/spec/template/spec/providerSpec/value/securityGroups/0/filters/0/values/-"),
+			Value: yamlpatch.NewNode(&securityGroupFilter),
+		},
+	}
+
+	// Apply patch to manifest
+	modifiedBytes, err = ops.Apply(manifestBytes)
+	if err != nil {
+		logger.WithError(err).Error("error applying patch")
+		return modifiedBytes, modified, err
+	}
+
+	if res := bytes.Compare(modifiedBytes, manifestBytes); res != 0 {
+		modified = true
+	}
```

Comment:
I'm adding a test for `"security group already configured"` which I would have assumed does not result in `modified = true` but it appears that if `Apply()` is successful the document will be modified. This is probably true unless your document matches exactly what `yamlpatch` would have output which won't occur in the wild?

Reply:
The way we're patching, the SG will get appended to the list even if it's already there, resulting in duplicate entries. (I don't know if that would break AWS.)

As you say, that shouldn't happen in real life because the manifest won't contain that SG initially, so it's probably not worth handling that case.

But what it does mean is that we probably don't need to bother with the Compare bit.

If you wanted to simplify further, we could make the return a `*[]byte` and `nil` means it wasn't modified. That might be stretching it though.